### ISSUE TITLE
#create? is called before fields are replaced?

### ISSUE THREAD

**Issue by NuckChorris** [venuu/jsonapi-authorization#25]
According to hummingbird-me/hummingbird#748 it seems that `#create?` is being called before the fields are replaced, which makes things like ownership checks fail.

For example, we use a `#create?` check to verify that `record.user_id === user.id`, to prevent users from impersonating other users, without implicitly adding that field.

I haven't looked too deeply into how you handle this, but in our previous solution we did the authorization check with [after_replace_fields](https://github.com/hummingbird-me/hummingbird/commit/723ee431221550e8fa5084f74e53cca89d5e1f35#diff-31800b00cba367352f0b5714ebbe706fL50), which worked great.

**Comment by valscion** [comment]
Yeah we're running all the authorizations in [`before` callbacks](https://github.com/venuu/jsonapi-authorization/blob/v0.6.1/lib/jsonapi/authorization/authorizing_operations_processor.rb#L6-L18).

In a sense, it makes sense that authorization is checked _before_ anything is applied to the database. That way we can stop the request at its tracks before any change is even tried to be applied.

This choice also means that the [authorizer class will only get the model class](https://github.com/venuu/jsonapi-authorization/blob/v0.6.1/lib/jsonapi/authorization/default_pundit_authorizer.rb#L102-L116) that is trying to be created, and we cannot query the fields on the underlying model as it isn't yet created even in memory.

I see your use case though, and it feels like we should have some way of supporting that. I'll think about it. Thanks for raising this issue!

**Comment by NuckChorris** [comment]
The way we did it, was after the fields are replaced by before they're persisted to the database. I can't imagine any downside to checking it at that point

**Comment by valscion** [comment]
I think some associations might be saved immediately when they're set, so replacing fields first might not be wise

**Comment by NuckChorris** [comment]
Really? I don't think ActiveRecord would do that, I'm pretty sure it just stores the changes when you `save` said, even for complex associations like HABTM or has-many-through

**Comment by valscion** [comment]
Yeah that actually would be true for a new record that hasn't been persisted yet. However, for an already persisted record, just setting association would trigger a database operation:

```
RAILS_ENV=test irb -r './spec/dummy/config/environment.rb'

article = Article.find_by(external_id: '2') # Does not have comments
comments = Comment.last(2)

# This will change the association even if I don't say `save`
article.comments = comments
```

It feels a bit flaky to rely on this functionality where assigning to a new record doesn't do database operations but assigning to a persisted record would change the database.

----

Eventually, it might even be that your use case would also be solved if we could get to some sort of resolution with #15. That way you could still just check for "user presence" on `create?` and then authorize the association assignment separately. I think that even if you would use Pundit with regular Rails controllers, you might want to go with this kind of code:

```ruby
class ArticlesController
  def create
    authorize Article, :create?
    Article.create #....
  end
end
```

and you would only be able to use the class instead of the record in the pundit policies, too.

**Comment by valscion** [comment]
I feel like this is actually a feature and not a bug. The use case behind this would also be solved if #15 was solved, so I'll close this now. For more reasoning, see https://github.com/hummingbird-me/hummingbird/pull/748#issuecomment-237833108

**Comment by valscion** [comment]
Please see #30 for a possible solution to this problem. Any input in there is greatly appreciated!
